### ISSUE TITLE
HasMany().WithOne().IsRequired() relation is not reflected by the changetracker before SaveChanges() is invoked

### ISSUE THREAD

**Issue by Tasteful** [dotnet/efcore#4220]
I have an entity `A` with a list of child entities `B` that is configured with an required hasmany-withone relationship and removing one of the child entities `B` from it's parent collection is not reflected by the changetracker before the `dbContext.SaveChange` method is invoked.

In my project I will use the information in the changetracker to fire application different application events depend on the entity and state combination and in some cases I need the data still in db to be able to fetch the data for application event.

Is this by design? Can I trigger to also get orphaned entries with state deleted before invoking the SaveChange?

```
info: BugTest.TestRunner[0]
      ChangeTracker.AutoDetectChangesEnabled: True
info: BugTest.TestRunner[0]
      Start dumping change tracker states: Add entities
      Name: BugTest.A id: 0b7801c1-60c3-475f-9982-8aa7cf2a2a55 state: Added
      Name: BugTest.B id: b6f65c34-9e45-4a51-958f-e5f25ab6d935 state: Added
      Name: BugTest.B id: 1548652a-4dc7-4373-8a8b-09534db32680 state: Added
      End dumping change tracker states: Add entities

info: BugTest.TestRunner[0]
      Invoke SaveChanges
info: Microsoft.Data.Entity.Storage.Internal.InMemoryStore[1]
      Saved 3 entities to in-memory store.
info: BugTest.TestRunner[0]
      Start dumping change tracker states: Before remove
      Name: BugTest.A id: 0b7801c1-60c3-475f-9982-8aa7cf2a2a55 state: Unchanged
      Name: BugTest.B id: b6f65c34-9e45-4a51-958f-e5f25ab6d935 state: Unchanged
      Name: BugTest.B id: 1548652a-4dc7-4373-8a8b-09534db32680 state: Unchanged
      End dumping change tracker states: Before remove

info: BugTest.TestRunner[0]
      Remove child entity.
info: BugTest.TestRunner[0]
      Start dumping change tracker states: After remove
      Name: BugTest.A id: 0b7801c1-60c3-475f-9982-8aa7cf2a2a55 state: Unchanged
      Name: BugTest.B id: b6f65c34-9e45-4a51-958f-e5f25ab6d935 state: Unchanged
      Name: BugTest.B id: 1548652a-4dc7-4373-8a8b-09534db32680 state: Unchanged
      End dumping change tracker states: After remove

info: BugTest.TestRunner[0]
      Invoke SaveChanges(false)
info: Microsoft.Data.Entity.Storage.Internal.InMemoryStore[1]
      Saved 1 entities to in-memory store.
info: BugTest.TestRunner[0]
      Start dumping change tracker states: After save
      Name: BugTest.A id: 0b7801c1-60c3-475f-9982-8aa7cf2a2a55 state: Unchanged
      Name: BugTest.B id: b6f65c34-9e45-4a51-958f-e5f25ab6d935 state: Deleted
      Name: BugTest.B id: 1548652a-4dc7-4373-8a8b-09534db32680 state: Unchanged
      End dumping change tracker states: After save
```

Repro with code https://github.com/Tasteful/bugs/tree/EF7-ChangeTracker and the program to run is executing the https://github.com/Tasteful/bugs/blob/EF7-ChangeTracker/src/BugTest/TestClasses.cs that contains all the relevant code.

**Comment by rowanmiller** [comment]
This is by-design that we now only cascade on `SaveChanges`. @ajcvickers is there any way to manually trigger the process before `SaveChanges`

**Comment by rowanmiller** [comment]
As mentioned, this is by design. But we did open https://github.com/aspnet/EntityFramework/issues/4290 to look at a way to eagerly apply cascade deletes in the future.

**Comment by Tasteful** [comment]
Thanks for the feedback. Do you think it will be included in or after 7.0.0 release?

**Comment by rowanmiller** [comment]
@Tasteful no, we won't have time to implement it before 7.0.0
